# NB_09 — Stage 9: Zero-Shot Qwen Baseline

This notebook evaluates the **base Qwen2.5-VL-7B-Instruct model without any
LoRA adapter** on the same 280-sample eval set used in NB_06.

**Why this matters:** NB_08 compared the fine-tuned Qwen against EasyOCR,
but that comparison has a built-in asymmetry — Qwen was trained on GPT labels,
so measuring both against GPT as reference favors Qwen by construction. A
fairer measure of what fine-tuning actually contributed is comparing the
fine-tuned model against the same base model with no training. Both are
measured against the same GPT reference, so any difference in CER/WER is
attributable purely to the LoRA fine-tuning.

**Expected result:** The base model should produce plausible Arabic text
(it is a strong multilingual VLM) but will not follow the JSON output format
reliably and will not be calibrated to AHTD's handwriting style. CER/WER
should be substantially worse than the fine-tuned checkpoint.

**Hardware:** A100 required (same as NB_06).  
**Prerequisites:** NB_00 (data setup). No adapter needed.

---
## Step 9.1 — Mount Drive and set paths

In [16]:
from google.colab import drive
drive.mount('/content/drive')

import os

PROJECT_ROOT = '/content/drive/MyDrive/CUD files/NLP/NLP_Arabic_HTR_Project'
EVAL_FILE    = f'{PROJECT_ROOT}/data/eval/eval.jsonl'
LOG_DIR      = f'{PROJECT_ROOT}/logs/run-1'
os.makedirs(LOG_DIR, exist_ok=True)

BASE_MODEL  = 'Qwen/Qwen2.5-VL-7B-Instruct'
MIN_PIXELS  = 4   * 28 * 28
MAX_PIXELS  = 128 * 28 * 28

print(f'PROJECT_ROOT : {PROJECT_ROOT}')
print(f'Base model   : {BASE_MODEL}')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
PROJECT_ROOT : /content/drive/MyDrive/CUD files/NLP/NLP_Arabic_HTR_Project
Base model   : Qwen/Qwen2.5-VL-7B-Instruct


---
## Step 9.2 — Install dependencies

In [2]:
import subprocess, sys, torch

subprocess.run([
    sys.executable, '-m', 'pip', 'install', '-q',
    'transformers>=4.40',
    'bitsandbytes>=0.43',
    'accelerate>=0.30',
    'qwen-vl-utils',
    'jiwer',
], check=True)

assert torch.cuda.is_available(), 'No GPU — switch runtime to A100.'
print(f'torch {torch.__version__}  |  GPU: {torch.cuda.get_device_name(0)}')
print('Dependencies installed.')

torch 2.10.0+cu128  |  GPU: NVIDIA A100-SXM4-40GB
Dependencies installed.


---
## Step 9.3 — Load the base model (no adapter)

The base model is loaded in 4-bit NF4 with the same quantization config used
in NB_05/06. No LoRA adapter is applied. This is the model as it exists
before any fine-tuning on AHTD.

In [3]:
import torch
from transformers import AutoProcessor, Qwen2_5_VLForConditionalGeneration, BitsAndBytesConfig

bnb_config = BitsAndBytesConfig(
    load_in_4bit              = True,
    bnb_4bit_quant_type       = 'nf4',
    bnb_4bit_use_double_quant = True,
    bnb_4bit_compute_dtype    = torch.bfloat16,
)

print(f'Loading {BASE_MODEL} (no adapter)...')
model = Qwen2_5_VLForConditionalGeneration.from_pretrained(
    BASE_MODEL,
    quantization_config = bnb_config,
    device_map          = 'auto',
    torch_dtype         = torch.bfloat16,
    attn_implementation = 'sdpa',
)
model.config.use_cache = True
model.eval()

processor = AutoProcessor.from_pretrained(
    BASE_MODEL,
    min_pixels = MIN_PIXELS,
    max_pixels = MAX_PIXELS,
)

print('Base model loaded. No LoRA adapter applied.')

Loading Qwen/Qwen2.5-VL-7B-Instruct (no adapter)...


config.json: 0.00B [00:00, ?B/s]

model.safetensors.index.json: 0.00B [00:00, ?B/s]

Fetching 5 files:   0%|          | 0/5 [00:00<?, ?it/s]

Loading weights:   0%|          | 0/729 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/216 [00:00<?, ?B/s]

preprocessor_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

chat_template.json: 0.00B [00:00, ?B/s]

The image processor of type `Qwen2VLImageProcessor` is now loaded as a fast processor by default, even if the model checkpoint was saved with a slow processor. This is a breaking change and may produce slightly different outputs. To continue using the slow processor, instantiate this class with `use_fast=False`. 


tokenizer_config.json: 0.00B [00:00, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Base model loaded. No LoRA adapter applied.


---
## Step 9.4 — Load eval samples and define helpers

Same image decoder and JSON parser used in NB_06, reproduced here so this
notebook is self-contained.

In [4]:
import json, base64, time
from io import BytesIO
from PIL import Image
from jiwer import cer as jiwer_cer, wer as jiwer_wer


def extract_images(messages):
    images = []
    for msg in messages:
        content = msg.get('content')
        if not isinstance(content, list):
            continue
        for item in content:
            t = item.get('type')
            if t == 'image_url':
                url = item['image_url']
                if isinstance(url, dict):
                    url = url.get('url', '')
                if url.startswith('data:'):
                    b64 = url.split(',', 1)[1]
                    images.append(Image.open(BytesIO(base64.b64decode(b64))).convert('RGB'))
    return images


def parse_output(raw_text):
    text = raw_text.strip()
    try:
        return json.loads(text)
    except json.JSONDecodeError:
        pass
    for suffix in ['"}"', ' "}', '"}', '"]}', '"}}']:
        try:
            return json.loads(text + suffix)
        except json.JSONDecodeError:
            continue
    return None


eval_samples = []
with open(EVAL_FILE) as f:
    for line in f:
        eval_samples.append(json.loads(line))

eos_token_id = processor.tokenizer.eos_token_id
pad_token_id = processor.tokenizer.pad_token_id or eos_token_id

print(f'Loaded {len(eval_samples)} eval samples.')

Loaded 280 eval samples.


---
## Step 9.5 — Run zero-shot inference

The generation setup is identical to NB_06: JSON prefix forcing, greedy
decoding, KV cache enabled, same max_new_tokens and repetition penalty.
The only difference is that no LoRA weights are loaded — the model is
responding purely from its pre-training.

In [5]:
results = []
json_prefix = '{"transcription":"'

print(f'Running zero-shot inference on {len(eval_samples)} samples...')

for i, sample in enumerate(eval_samples):
    messages      = sample['messages']
    input_messages = messages[:-1]   # system + user only
    expected      = messages[-1]['content']

    text = processor.apply_chat_template(
        input_messages, tokenize=False, add_generation_prompt=True
    ) + json_prefix

    image_inputs = extract_images(input_messages)

    inputs = processor(
        text   = [text],
        images = image_inputs if image_inputs else None,
        padding       = True,
        return_tensors = 'pt',
    ).to(model.device)

    t0 = time.time()
    with torch.inference_mode():
        output_ids = model.generate(
            **inputs,
            max_new_tokens       = 128,
            do_sample            = False,
            use_cache            = True,
            repetition_penalty   = 1.1,
            no_repeat_ngram_size = 6,
            eos_token_id         = eos_token_id,
            pad_token_id         = pad_token_id,
        )
    elapsed = time.time() - t0

    generated = processor.decode(
        output_ids[0][inputs['input_ids'].shape[1]:],
        skip_special_tokens=True
    )
    full_output  = json_prefix + generated
    pred_parsed  = parse_output(full_output)
    ref_parsed   = parse_output(expected)

    pred_text = pred_parsed.get('transcription', '').strip() if pred_parsed else ''
    ref_text  = ref_parsed.get('transcription',  '').strip() if ref_parsed  else ''

    cer_val   = float(jiwer_cer(ref_text, pred_text)) if ref_text else 0.0
    wer_val   = float(jiwer_wer(ref_text, pred_text)) if ref_text.strip() else 0.0
    exact     = pred_text.strip() == ref_text.strip()

    results.append({
        'cer':            cer_val,
        'wer':            wer_val,
        'exact_match':    exact,
        'valid_json':     pred_parsed is not None,
        'inference_time': elapsed,
        'prediction':     pred_text,
        'reference':      ref_text,
    })

    if (i + 1) % 20 == 0:
        print(f'  [{i+1}/280] elapsed: {sum(r["inference_time"] for r in results):.1f}s')

print('\nZero-shot inference complete.')

The following generation flags are not valid and may be ignored: ['temperature']. Set `TRANSFORMERS_VERBOSITY=info` for more details.


Running zero-shot inference on 280 samples...
  [20/280] elapsed: 52.0s
  [40/280] elapsed: 111.5s
  [60/280] elapsed: 157.1s
  [80/280] elapsed: 216.8s
  [100/280] elapsed: 261.1s
  [120/280] elapsed: 304.7s
  [140/280] elapsed: 358.6s
  [160/280] elapsed: 397.8s
  [180/280] elapsed: 442.3s
  [200/280] elapsed: 486.1s
  [220/280] elapsed: 529.4s
  [240/280] elapsed: 582.9s
  [260/280] elapsed: 620.5s
  [280/280] elapsed: 674.9s

Zero-shot inference complete.


---
## Step 9.6 — Compute and save metrics

In [13]:
import json, os

n = len(results)

avg_cer     = sum(r['cer']            for r in results) / n
avg_wer     = sum(r['wer']            for r in results) / n
exact_rate  = sum(r['exact_match']    for r in results) / n * 100
json_rate   = sum(r['valid_json']     for r in results) / n * 100
avg_time    = sum(r['inference_time'] for r in results) / n

print('=== Zero-Shot Qwen2.5-VL-7B Results ===')
print(f'  Num samples      : {n}')
print(f'  CER              : {avg_cer:.4f}')
print(f'  WER              : {avg_wer:.4f}')
print(f'  Exact match rate : {exact_rate:.2f}%')
print(f'  Valid JSON rate  : {json_rate:.2f}%')
print(f'  Avg infer time   : {avg_time:.2f} s/sample')

summary = {
    'model':              BASE_MODEL,
    'adapter':            'none (zero-shot)',
    'num_samples':        n,
    'avg_cer':            round(avg_cer,    4),
    'avg_wer':            round(avg_wer,    4),
    'exact_match_rate':   round(exact_rate, 2),
    'valid_json_rate':    round(json_rate,  2),
    'avg_inference_time': round(avg_time,   2),
    'per_sample': [
        {
            'reference':      r['reference'],
            'prediction':     r['prediction'],
            'cer':            r['cer'],
            'wer':            r['wer'],
            'exact_match':    r['exact_match'],
            'valid_json':     r['valid_json'],
            'inference_time': r['inference_time'],
        }
        for r in results
    ],
}

out_path = f'{LOG_DIR}/eval_results_zero_shot.json'
with open(out_path, 'w', encoding='utf-8') as f:
    json.dump(summary, f, indent=2, ensure_ascii=False)

print(f'\nResults saved to {out_path}')
print(f'per_sample entries saved: {len(summary["per_sample"])}')

=== Zero-Shot Qwen2.5-VL-7B Results ===
  Num samples      : 280
  CER              : 0.4639
  WER              : 0.7755
  Exact match rate : 0.71%
  Valid JSON rate  : 97.50%
  Avg infer time   : 2.41 s/sample

Results saved to /content/drive/MyDrive/CUD files/NLP/NLP_Arabic_HTR_Project/logs/run-2/eval_results_zero_shot.json
per_sample entries saved: 280


---
## Step 9.7 — Three-way comparison

Load results from NB_06 (fine-tuned Qwen) and NB_08 (EasyOCR) alongside
the zero-shot numbers to produce the complete baseline comparison table.

This is the primary result of Stage 9: the gap between zero-shot and
fine-tuned Qwen measures exactly what the knowledge distillation training
contributed, free of the reference-bias problem discussed in NB_08.

In [15]:
import json

with open(f'{PROJECT_ROOT}/logs/run-1/eval_results_checkpoint-1120.json') as f:
    finetuned = json.load(f)

with open(f'{PROJECT_ROOT}/logs/stage7/easyocr_results.json') as f:
    easyocr = json.load(f)

print('=== Complete Baseline Comparison (280-sample eval set) ===\n')
print(f'{"Model":<45} {"CER":>8} {"WER":>8} {"Valid JSON":>12}')
print('-' * 78)
print(f'{"Qwen2.5-VL-7B + LoRA (checkpoint-1120)":<45} {finetuned["avg_cer"]:>8.4f} {finetuned["avg_wer"]:>8.4f} {str(finetuned["valid_json_rate"])+"%":>12}')
print(f'{"Qwen2.5-VL-7B zero-shot (no fine-tuning)":<45} {avg_cer:>8.4f} {avg_wer:>8.4f} {str(round(json_rate,1))+"%":>12}')
print(f'{"EasyOCR (Arabic, off-the-shelf)":<45} {easyocr["cer"]:>8.4f} {easyocr["wer"]:>8.4f} {"N/A":>12}')
print()
print(f'Fine-tuning improvement over zero-shot:')
print(f'  CER reduction: {avg_cer - finetuned["avg_cer"]:.4f} ({(avg_cer - finetuned["avg_cer"]) / avg_cer * 100:.1f}%)')
print(f'  WER reduction: {avg_wer - finetuned["avg_wer"]:.4f} ({(avg_wer - finetuned["avg_wer"]) / avg_wer * 100:.1f}%)')
print()
print('Note: all models evaluated against GPT-5.4-mini pseudo-labels.')
print('EasyOCR has no JSON output — valid JSON rate is not applicable.')

=== Complete Baseline Comparison (280-sample eval set) ===

Model                                              CER      WER   Valid JSON
------------------------------------------------------------------------------
Qwen2.5-VL-7B + LoRA (checkpoint-1120)          0.3283   0.6554       92.14%
Qwen2.5-VL-7B zero-shot (no fine-tuning)        0.4639   0.7755        97.5%
EasyOCR (Arabic, off-the-shelf)                 0.4764   1.0582          N/A

Fine-tuning improvement over zero-shot:
  CER reduction: 0.1356 (29.2%)
  WER reduction: 0.1201 (15.5%)

Note: all models evaluated against GPT-5.4-mini pseudo-labels.
EasyOCR has no JSON output — valid JSON rate is not applicable.


# review results summary

These results are clean and tell a coherent story across all three models.

**Zero-shot Qwen is already a strong model**

CER of 0.464 and WER of 0.776 without any fine-tuning is a notable result on its own. Qwen2.5-VL-7B was trained on a massive multilingual corpus that includes Arabic text in images, so it arrives with genuine Arabic reading capability. The 97.5% valid JSON rate is also striking — the base model follows the output format instruction almost perfectly even without task-specific training, which reflects the quality of its instruction tuning. This is the baseline you are actually competing against when you do knowledge distillation: not a random model, but an already-capable one.

**Fine-tuning improved CER by 29%, WER by 15%**

The LoRA adapter trained on 1,120 GPT-labeled samples reduced CER from 0.464 to 0.328 and WER from 0.776 to 0.655. These are meaningful, consistent improvements across both metrics. The CER improvement is larger than the WER improvement, which makes sense: fine-tuning on handwriting samples teaches the model to read individual Arabic characters more accurately, and CER captures character-level fidelity more directly than WER does.

**The valid JSON rate inversion is interesting**

The fine-tuned model has a lower valid JSON rate (92.1%) than the zero-shot base (97.5%). This is a side effect of how training was conducted: the full sequence including system prompt and image tokens was included in the label, so the model was not exclusively trained to produce the JSON wrapper. Some fraction of outputs drifted from the strict format. The base model, by contrast, was instruction-tuned specifically to follow output format instructions reliably. For the paper, this is worth noting: fine-tuning improved transcription quality at a small cost to output format reliability.

**EasyOCR is the weakest of the three**

EasyOCR's CER (0.476) is barely higher than zero-shot Qwen (0.464), but its WER (1.058) is dramatically worse. This confirms something the qualitative samples in NB_08 already suggested: EasyOCR produces roughly the right characters in isolated bursts but fails at word-level segmentation on Arabic paragraph images. Arabic is a cursive script and word boundaries depend heavily on context — a statistical OCR engine struggles with this in a way that a large VLM does not.

**What this means for the paper**

The three-way table now tells a clean progression: EasyOCR ≈ zero-shot Qwen on CER, but fine-tuned Qwen is clearly better than both. The fine-tuning contribution is real and quantifiable at roughly 29% CER reduction. The comparison is now methodologically sound: the primary baseline is zero-shot Qwen (same model, same reference, no training advantage), and EasyOCR serves as a classical OCR reference point. The ground-truth limitation applies equally to all three models, so relative comparisons are valid even if absolute values are not.